DragaliaAPI/Controllers/Dragalia/AbilityCrestController.cs

```diff
@@ -152,4 +156,88 @@ public async Task<DragaliaResult> ResetPlusCount(AbilityCrestResetPlusCountReque
         UpdateDataList updateDataList = await this.updateDataService.SaveChangesAsync();
         return Ok(new AbilityCrestResetPlusCountData() { update_data_list = updateDataList });
     }
+
+    [Route("get_ability_crest_set_list")]
+    [HttpPost]
+    public async Task<DragaliaResult> GetAbilityCrestSetList(
+        AbilityCrestGetAbilityCrestSetListRequest request
+    )
+    {
+        List<DbAbilityCrestSet> dbAbilityCrestSets = await abilityCrestRepository.AbilityCrestSets
+            .OrderBy(x => x.AbilityCrestSetNo)
+            .ToListAsync();
+
+        int index = 0;
+
+        IEnumerable<AbilityCrestSetList> abilityCrestSetList = Enumerable
+            .Range(1, 54)
+            .Select(
+                x =>
+                    index < dbAbilityCrestSets.Count()
+                    && dbAbilityCrestSets[index].AbilityCrestSetNo == x
+                        ? dbAbilityCrestSets[index++]
+                        : new DbAbilityCrestSet(DeviceAccountId, x)
+            )
+            .Select(mapper.Map<AbilityCrestSetList>)
+            .ToArray();
+
+        return Ok(
+            new AbilityCrestGetAbilityCrestSetListData()
+            {
+                ability_crest_set_list = abilityCrestSetList
+            }
+        );
+    }
+
+    [Route("set_ability_crest_set")]
+    [HttpPost]
+    public async Task<DragaliaResult> SetAbilityCrestSet(
+        AbilityCrestSetAbilityCrestSetRequest request
+    )
+    {
+        if (request.ability_crest_set_no is <= 0 or > 54)
+        {
+            this.logger.LogError("Invalid ability crest no", request.ability_crest_set_no);
+            return this.Code(ResultCode.CommonInvalidArgument);
+        }
+
+        DbAbilityCrestSet newAbilityCrestSet = mapper.Map<DbAbilityCrestSet>(request);
+        newAbilityCrestSet.DeviceAccountId = DeviceAccountId;
```

Comment:
This also doesn't respect user impersonation and in this case it does matter, and as such I'd recommend making IAbilityCrestRepository::AddOrUpdateSet set the DeviceAccountId of the passed-in db record so it's always assigned to the correct user.